### ISSUE TITLE
extend FFI::Library doesn't extend some methods into module

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1529]
There is no issue under MRI. After initializing the module, `TestModule': undefined method 'callback' for TestModule:Module (NoMethodError)` exception is occurred. The code below tries to call this method https://www.rubydoc.info/github/ffi/ffi/FFI/Library#callback-instance_method

```ruby
require 'ffi'

module TestModule
  extend FFI::Library

  callback :g_log_func, [:string, :int, :string, :pointer], :void
end

TestModule
```

This is a small reproducible example of issue comes from https://github.com/libvips/ruby-vips/blob/428d5c106fb4a64521159c6b0c4f9391eea2eba6/lib/vips.rb#L36

Error:

```
ruby-ffi.rb:6:in `TestModule': undefined method `callback' for TestModule:Module (NoMethodError)
	from ruby-ffi.rb:3:in `<main>'
```

**Comment by deepj** [comment]
It seems to be similar to #1524

**Comment by eregon** [comment]
Our current implementation of FFI (using Truffle NFI) is not complete, so this means we did not implement that method yet.
It's an open question currently whether we can and how to reuse some of the upstream FFI code (we likely wouldn't want a second `libffi` copy, so using Truffle NFI would be better).

**Comment by eregon** [comment]
@deepj The `ffi` gem is installed, but it no-ops when `require`-d: https://github.com/ffi/ffi/blob/0ee4f7931bb56c7d0e11ef66c51711cb59a80bc4/lib/ffi.rb#L12

**Comment by deepj** [comment]
@eregon I see, got it.

**Comment by eregon** [comment]
This is also an issue for `ethon` (#1524) and `sassc`.
Related issue in upstream FFI: https://github.com/ffi/ffi/issues/660
In short, I plan to do a pure-Ruby implementation of FFI using Truffle NFI as backend.

**Comment by deepj** [comment]
@eregon @chrisseaton if there would be a chance to move FFI compability forward it could help to run more native gems on TruffleRuby. RC13/14?

**Comment by chrisseaton** [comment]
I think we're unlikely to fix this one as fast as the other priority issues, I'm afraid. It's going to be a bit more work.

**Comment by chrisseaton** [comment]
I'm going to remove the *priority* tag on this, as we aren't going to be able to solve it in the next month or so, which is what we're aiming at with priority issues now.

**Comment by eregon** [comment]
I fixed this in 59fb511e431b4795580ce7d191b84f0199351012 and added complete support for the FFI, with a pure-Ruby implementation using Truffle NFI.
It will be in the next release (1.0.0-rc16).
Thank you for the report.
